libbeat/autodiscover/providers/kubernetes/config.go

```diff
@@ -102,6 +119,41 @@ func (c *Config) Validate() error {
 	if c.Scope != "node" && c.Scope != "cluster" {
 		return fmt.Errorf("invalid `scope` configured. supported values are `node` and `cluster`")
 	}
+
+	if c.Sharding.Count != 0 {
+		if c.Scope == "node" {
+			logp.L().Warnf("can not set sharding.count to `%d` when scope to `cluster`", c.Sharding.Count)
+			c.Sharding.Count = 0
+			c.Sharding.Instance = -1
+		} else {
+			// Ensure that if a user decides to run cluster mode sharded Beats, then each instance and its replica do leader election.
+			if c.Unique && c.LeaderLease != "" {
+				c.LeaderLease = fmt.Sprintf("%s-%d", c.LeaderLease, c.Sharding.Instance)
```

Comment:
Something that could be done is that if both `Unique` and `Sharding` are enabled, `sharding.instance` is ignored, and running instances auto-organize using leases.

E.g. we could have five beats running with this config:
```
unique: true
leader_lease: "some-lease"
sharding.count: 4
```
All instances would try to get the lease for `some-lease-0` to `some-lease-3`. Four of them would take a lease, and will start watching the corresponding shard. The fifth one would be waiting. In case one of the lease is released because a pod dies or a rolling release is happening, the fifth instance could take over.

Reply:
@ChrsMark this will create a different lease per shard. if we have mb-0, mb-1 and mb-2 we would have to the workload sharded across all three.

we would additionally create mb-back-0, mb-back-1, mb-back-2 as standbys for the above ones which would leader elect with the corresponding active shard. hence why we have done this.

Reply:
@jsoriano that is a very interesting approach. my thought was to either run this as follows:

* run 2 stateful sets that use the same `leader_lease` and the 0th instance of both would elect to provide fault tolerance especially roll outs
* have N deployments with 2 replicas so that each deployment can act as instance 0, 1, 2 etc and would have leader election across the replicas. during rollouts the leader election would replace the active instance. 


with your approach, i have 2 questions:
* how do we ensure deterministic behavior if the workloads can move around between the pods? i.e, it becomes very difficult to debug if scraping is not happening. in the current approach, i can take the pod uuid, generate the hash and mod it by count to figure out which instance is supposed to monitor said pod. 
* how do we provide HA? during a code rollout we would blank out `1/count*100 %` of the entire cluster.

Reply:
> * how do we ensure deterministic behavior if the workloads can move around between the pods? i.e, it becomes very difficult to debug if scraping is not happening. in the current approach, i can take the pod uuid, generate the hash and mod it by count to figure out which instance is supposed to monitor said pod.
> 

You would do the same, but to see what instance is supposed to monitor the pod you would check the lease object. The holder should set a good identity in the lease lock config.

> * how do we provide HA? during a code rollout we would blank out `1/count*100 %` of the entire cluster.

In my example I was deploying 5 beats, but setting the `sharding.count` to 4, so there would be a spare instance. During a rollout the spare instances would take the released locks, providing the HA. All the instances would be competing for the N locks, but each one would take one at most.
There would be a single deployment (or statefulset) instead of N plus the stateful sets, making it easier to maintain. And there would be also less inactive instances.
A code rollout should be done following a rolling update strategy with `maxUnavailable` set to a number equal or less to the number of spare instances (to 1 in my example).

The problem I see now with both approaches is that the data directory won't be shared, though I don't think it is used for the use cases that would be covered by this feature.